Source/JavaScriptCore/builtins/SetPrototype.js

```diff
@@ -44,3 +44,254 @@ function forEach(callback /*, thisArg */)
         callback.@call(thisArg, key, key, this);
     } while (true);
 }
+
+function union(other)
+{
+    "use strict";
+
+    if (!@isSet(this))
+        @throwTypeError("Set operation called on non-Set object");
+
+    // Get Set Record
+    if (!@isObject(other))
+        @throwTypeError("Set.prototype.union expects first parameter to be an object");
+    var size = @toNumber(other.size);
+    // size is NaN
+    if (size !== size)
+        @throwTypeError("Set.prototype.union expects other.size to be a non-NaN number");
+
+    var has = other.has;
+    if (!@isCallable(has))
+        @throwTypeError("Set.prototype.union expects other.has to be callable");
+
+    var keys = other.keys;
+    if (!@isCallable(keys))
+        @throwTypeError("Set.prototype.union expects other.keys to be callable");
+
+    var iterator = keys.@call(other, keys);
+    var wrapper = {
+        @@iterator: function () { return iterator; }
+    };
+
+    var result = @setClone(this);
+    for (var key of wrapper)
+        result.@add(key);
+
+    return result;
+}
+
+function intersection(other)
+{
+    "use strict";
+
+    if (!@isSet(this))
+        @throwTypeError("Set operation called on non-Set object");
+
+    // Get Set Record
+    if (!@isObject(other))
+        @throwTypeError("Set.prototype.intersection expects first parameter to be an object");
+    var size = @toNumber(other.size);
+    // size is NaN
+    if (size !== size)
+        @throwTypeError("Set.prototype.intersection expects other.size to be a non-NaN number");
+
+    var has = other.has;
+    if (!@isCallable(has))
+        @throwTypeError("Set.prototype.intersection expects other.has to be callable");
+
+    var keys = other.keys;
+    if (!@isCallable(keys))
+        @throwTypeError("Set.prototype.intersection expects other.keys to be callable");
+
+    var result = new @Set();
+    if (this.@size <= size) {
+        var bucket = @setBucketHead(this);
+
+        do {
+            bucket = @setBucketNext(bucket);
+            if (bucket === @sentinelSetBucket)
+                break;
+            var key = @setBucketKey(bucket);
+            if (has.@call(other, key))
+                result.@add(key);
+        } while (true);
+    } else {
```

Comment:
Probably, we should have a Set fast path for other case. (iterating things like Set.forEach).

Reply:
it would be really cool if there was a way for C++-implemented classes to expose a fast path for iterators more generally. Like if the Iteratable object had an optional private property to call a function that returns the rest of the results as a JSArray or JSObject. `Array.from()`, `Object.fromEntries()`,  `[...spread]`, could use that automatically. Maybe a generic template with C++ iterators could be used to generate the fast path so that consumers in WebCore/elsewhere wouldn't need to bother with it

This would speed up iterators in Headers, the FileSystem Access API anywhere else
```js
    const headers = new Headers({
      "Last-Modified": new Date().toUTCString(),
      ETag: `W/abcdef`,
    });
   
   const pairs = [...headers.entries()]
   const keys = [...headers.keys()]
   const values = [...headers.keys()]
```

Reply:
> Probably, we should have a Set fast path for other case. (iterating things like Set.forEach).

Iterating via the wrapper should be just as fast in the DFG/FTL because we should inline all the calls (maybe a bit worse in the DFG because you have to allocate the wrapper).